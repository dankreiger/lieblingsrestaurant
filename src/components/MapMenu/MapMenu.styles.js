import styled from 'styled-components';
import { Col } from 'reactstrap';

/* this code is here for jest testing and component readability */
export const NavigationTogglerCol = styled(Col)`
  display: flex;
`;

export const MapSearchInputCol = styled(Col)``;
export const AddressTypeSelectCol = styled(Col)``;
