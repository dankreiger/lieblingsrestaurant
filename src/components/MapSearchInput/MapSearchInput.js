/*global google */

import React, { useRef, useState } from 'react';
import { array, func, object, shape } from 'prop-types';
import { BERLIN } from 'constants/index';
import { SearchInput, StyledPopover } from './MapSearchInput.styles';
import { PopoverBody } from 'reactstrap';
import idx from 'idx';

const MapSearchInput = ({ currentMapInfo, handleMapInstance, places }) => {
  const [showSuggestions, setSuggestionsVisibility] = useState(false);
  const [isRestaurant, setIsRestaurantStatus] = useState(true);
  const [lastFavoritedItemExists, setExistenceOfLastFavoritedItem] = useState(
    false
  );
  const inputContainer = useRef(null);

  const handleFocus = () => {
    setSuggestionsVisibility(false);
    setExistenceOfLastFavoritedItem(false);
    inputContainer.current.clear();
    if (!isRestaurant) {
      setIsRestaurantStatus(true);
    }
    // prevent suggestions dropdown from opening abruptly and unnecessarily
    // send a PR to Geosuggest
    setTimeout(() => {
      setSuggestionsVisibility(true);
    }, 1000);
  };

  const handleSuggestSelect = newPlace => {
    if (!newPlace) {
      return;
    }
    const { map, maps } = currentMapInfo;

    if (idx(newPlace, _ => _.gmaps.types.length)) {
      const newPlaceIsRestaurant = newPlace.gmaps.types.includes('restaurant');
      setIsRestaurantStatus(newPlaceIsRestaurant);
      if (!newPlaceIsRestaurant) {
        return;
      } else {
        if (!places.some(place => place.placeId === newPlace.placeId)) {
          setExistenceOfLastFavoritedItem(false);
          handleMapInstance(map, maps, [...places, newPlace]);
        } else {
          setExistenceOfLastFavoritedItem(true);
        }
      }
    }
  };
  return (
    <>
      <SearchInput
        ref={inputContainer}
        onFocus={handleFocus}
        placeholder="Search for a restaurant..."
        onSuggestSelect={handleSuggestSelect}
        location={new google.maps.LatLng(BERLIN.lat, BERLIN.lng)}
        radius={2000}
        types={['establishment']}
        errorExists={!isRestaurant || lastFavoritedItemExists}
        showSuggestions={showSuggestions}
        id="invalidSelectionPopup"
      />
      <StyledPopover
        placement="bottom"
        isOpen={!isRestaurant || lastFavoritedItemExists}
        target="invalidSelectionPopup"
      >
        <PopoverBody>
          {lastFavoritedItemExists
            ? 'You have already selected this restaurant!'
            : 'Sorry this is not a restaurant. Please try your search again.'}
        </PopoverBody>
      </StyledPopover>
    </>
  );
};

MapSearchInput.propTypes = {
  places: array,
  currentMapInfo: shape({
    map: object,
    maps: object
  }),
  handleMapInstance: func
};

export default MapSearchInput;
