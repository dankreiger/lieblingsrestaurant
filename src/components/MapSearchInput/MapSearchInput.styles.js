import styled from 'styled-components';
import Geosuggest from 'react-geosuggest';
import { Popover } from 'reactstrap';
import { DANGER_COLOR } from '../../constants';
import { ACTIVE_NAV_SHADOW, NAV_SHADOW } from '../../constants';

export const SearchInput = styled(Geosuggest)`
  &.geosuggest {
    font-size: 18px;
    font-size: 1rem;
    text-align: left;
    margin: 0 auto;
    width: 100vw;
    left: 0;
    position: absolute;
  }
  .geosuggest__suggests-wrapper {
    display: ${({ showSuggestions }) => (showSuggestions ? 'block' : 'none')};
  }
  .geosuggest__input {
    width: 100%;
    border: 2px solid
      ${({ errorExists }) => (errorExists ? DANGER_COLOR : 'transparent')};
    box-shadow: ${NAV_SHADOW};
    padding: 0.5em 1em;
    transition: border 0.2s, box-shadow 0.2s;
    &:focus {
      border-color: #267dc0;
      box-shadow: ${ACTIVE_NAV_SHADOW};
    }
    /* &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0px 1000px white inset;
    } */
  }
  .geosuggest__suggests {
    position: absolute;
    top: 100%;
    left: 0;
    right: 0;
    max-height: 25em;
    padding: 0;
    margin-top: -1px;
    background: #fff;
    border: 2px solid #267dc0;
    border-top-width: 0;
    overflow-x: hidden;
    overflow-y: auto;
    list-style: none;
    z-index: 5;
    -webkit-transition: max-height 0.2s, border 0.2s;
    transition: max-height 0.2s, border 0.2s;
  }
  .geosuggest__suggests--hidden {
    max-height: 0;
    overflow: hidden;
    border-width: 0;
  }

  /**
 * A geosuggest item
 */
  .geosuggest__item {
    font-size: 18px;
    font-size: 1rem;
    padding: 0.5em 0.65em;
    cursor: pointer;
    &:hover,
    &:focus {
      background: #f5f5f5;
    }
  }
  .geosuggest__item--active {
    background: #267dc0;
    color: #fff;
    &:hover,
    &:focus {
      background: #ccc;
    }
  }

  .geosuggest__item__matched-text {
    font-weight: bold;
  }
`;

export const StyledPopover = styled(Popover)`
  .popover-inner {
    border: 1px solid ${DANGER_COLOR};
  }
  .arrow {
    &:before {
      border-bottom-color: ${DANGER_COLOR} !important;
    }
  }
`;
