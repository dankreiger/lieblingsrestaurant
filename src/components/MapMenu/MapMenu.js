import React from 'react';
import { array, func, object } from 'prop-types';
import { Col, FormGroup, Input, Row } from 'reactstrap';
import MapSearchInput from 'components/MapSearchInput/MapSearchInput';

const MapMenu = ({ currentMapInfo, handleMapInstance, places }) => {
  return (
    <>
      <Row noGutters>
        <Col>
          <MapSearchInput
            handleMapInstance={handleMapInstance}
            places={places}
            currentMapInfo={currentMapInfo}
          />
        </Col>
        <Col>
          <FormGroup>
            <Input
              type="select"
              name="addressTypeFilter"
              id="addressTypeFilter"
            >
              <option>Establishments</option>
              <option>All Addresses</option>
            </Input>
          </FormGroup>
        </Col>
      </Row>
    </>
  );
};

MapMenu.propTypes = {
  currentMapInfo: object,
  handleMapInstance: func,
  places: array
};

export default MapMenu;
