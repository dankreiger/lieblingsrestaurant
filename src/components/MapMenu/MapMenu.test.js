import React from 'react';
import { shallow } from 'enzyme';
import MapMenu from './MapMenu';
import MapSearchInput from 'components/MapSearchInput/MapSearchInput';
import { FormGroup, Input } from 'reactstrap';

describe('App', () => {
  let mapMenuComponent;

  beforeEach(() => {
    mapMenuComponent = shallow(<MapMenu />);
  });

  describe('rendering', () => {
    test('renders as expected', () => {
      expect(mapMenuComponent.length).toBe(1);
      expect(mapMenuComponent).toMatchSnapshot();
    });
  });

  describe('structure', () => {
    test('renders 1 <MapSearchInput> component', () => {
      expect(mapMenuComponent.find(MapSearchInput).length).toBe(1);
    });
    test('renders 1 <FormGroup> component with a <Input> select menu component', () => {
      const formGroup = mapMenuComponent.find(FormGroup);
      expect(formGroup.length).toBe(1);

      expect(formGroup.find(Input).length).toBe(1);
      expect(formGroup.find(Input).props().type).toBe('select');
      expect(formGroup.find(Input).props().name).toBe('addressTypeFilter');
    });
  });
});
