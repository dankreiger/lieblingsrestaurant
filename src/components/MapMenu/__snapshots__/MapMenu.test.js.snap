// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MapMenu />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Row
        noGutters={true}
        tag="div"
      >
        <Col
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <MapSearchInput />
        </Col>
        <Col
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <FormGroup
            tag="div"
          >
            <Input
              id="addressTypeFilter"
              name="addressTypeFilter"
              type="select"
            >
              <option>
                Establishments
              </option>
              <option>
                All Addresses
              </option>
            </Input>
          </FormGroup>
        </Col>
      </Row>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Col
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <MapSearchInput />
          </Col>,
          <Col
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <FormGroup
              tag="div"
            >
              <Input
                id="addressTypeFilter"
                name="addressTypeFilter"
                type="select"
              >
                <option>
                  Establishments
                </option>
                <option>
                  All Addresses
                </option>
              </Input>
            </FormGroup>
          </Col>,
        ],
        "noGutters": true,
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <MapSearchInput />,
            "tag": "div",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "currentMapInfo": undefined,
              "handleMapInstance": undefined,
              "places": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <FormGroup
              tag="div"
            >
              <Input
                id="addressTypeFilter"
                name="addressTypeFilter"
                type="select"
              >
                <option>
                  Establishments
                </option>
                <option>
                  All Addresses
                </option>
              </Input>
            </FormGroup>,
            "tag": "div",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Input
                id="addressTypeFilter"
                name="addressTypeFilter"
                type="select"
              >
                <option>
                  Establishments
                </option>
                <option>
                  All Addresses
                </option>
              </Input>,
              "tag": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <option>
                    Establishments
                  </option>,
                  <option>
                    All Addresses
                  </option>,
                ],
                "id": "addressTypeFilter",
                "name": "addressTypeFilter",
                "type": "select",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Establishments",
                  },
                  "ref": null,
                  "rendered": "Establishments",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "All Addresses",
                  },
                  "ref": null,
                  "rendered": "All Addresses",
                  "type": "option",
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Row
          noGutters={true}
          tag="div"
        >
          <Col
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <MapSearchInput />
          </Col>
          <Col
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <FormGroup
              tag="div"
            >
              <Input
                id="addressTypeFilter"
                name="addressTypeFilter"
                type="select"
              >
                <option>
                  Establishments
                </option>
                <option>
                  All Addresses
                </option>
              </Input>
            </FormGroup>
          </Col>
        </Row>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Col
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <MapSearchInput />
            </Col>,
            <Col
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <FormGroup
                tag="div"
              >
                <Input
                  id="addressTypeFilter"
                  name="addressTypeFilter"
                  type="select"
                >
                  <option>
                    Establishments
                  </option>
                  <option>
                    All Addresses
                  </option>
                </Input>
              </FormGroup>
            </Col>,
          ],
          "noGutters": true,
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MapSearchInput />,
              "tag": "div",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentMapInfo": undefined,
                "handleMapInstance": undefined,
                "places": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <FormGroup
                tag="div"
              >
                <Input
                  id="addressTypeFilter"
                  name="addressTypeFilter"
                  type="select"
                >
                  <option>
                    Establishments
                  </option>
                  <option>
                    All Addresses
                  </option>
                </Input>
              </FormGroup>,
              "tag": "div",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Input
                  id="addressTypeFilter"
                  name="addressTypeFilter"
                  type="select"
                >
                  <option>
                    Establishments
                  </option>
                  <option>
                    All Addresses
                  </option>
                </Input>,
                "tag": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    <option>
                      Establishments
                    </option>,
                    <option>
                      All Addresses
                    </option>,
                  ],
                  "id": "addressTypeFilter",
                  "name": "addressTypeFilter",
                  "type": "select",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Establishments",
                    },
                    "ref": null,
                    "rendered": "Establishments",
                    "type": "option",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "All Addresses",
                    },
                    "ref": null,
                    "rendered": "All Addresses",
                    "type": "option",
                  },
                ],
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
